BoomBoxEx4.c

```diff
@@ -0,0 +1,271 @@
+  }
+  delay(noteDuration);
+}
+
+// Plays a song in the form of a string. The format of the string is based on the
+// PLAY statement in IBM ADVANCED BASIC: 
+//   https://archive.org/details/IBMBASICAV1.10Manual/page/n315/mode/2up
+//
+// Implements:
+//   A to G with optional #, +, or -:
+//       Plays the indicated note in the current octave. A number sign (#) 
+//       or plus sign (+) afterwards indicates a sharp, a minus sign (-)
+//       indicates a flat. The #, +, or - is not allowed unless it corresponds
+//       to a black key on a piano. For example, B# is an invalid note.
+//
+//   O n: 
+//       Octave. Sets the current octave for the following notes. There are
+//       7 octaves, numbered 0 to 6. Each octave goes from C to B. Octave 3
+//       starts with middle C. Octave 4 is the default octave.
+//
+//   L n:
+//       Sets the length of the following notes. The actual note length is 1/n.
+//       n may range from 1 to 64. The following table may help explain this:
+//
+//       Length   Equivalent
+//         L1     whole note
+//         L2     half note
+//         L3     one of a triplet of three half notes (1/3 of a 4 beat measure)
+//         L4     quarter note
+//         L5     one of a quintuplet (1/5 of a measure)
+//         L6     one of a quarter note triplet
+//         .
+//         .
+//         .
+//         L64    sixty-fourth note
+//
+//       The length may also follow the note when you want to change the length
+//       only for the note. For example, A16 is equivalent to LIGA.
+//
+//   P n:
+//       Pause (rest). n may range from 1 to 64, and figures the length of the 
+//       pause in the same way as L (length).
+//
+//   (dot or period):
+//       After a note, causes the note to be played as a dotted note. That is,
+//       its length is multiplied by 3/2. Dots may also appear after a pause (P)
+//       to scale the pause length in the same way.
+//
+//   T n:
+//       Tempo. Sets the number of quarter notes in a minute. n may range from
+//       32 to 255. The default is 120.
+//
+void playSong(char *songStr) {
+  int tempo = 120;
+  int length = 4;
+  int octave = 4;
+  int note = -1;
+  
+  char *p = songStr;
+  while (*p != '\0') {
+    // skip whitespace
+    while (*p == ' ' || *p == ',') {
+      ++p;
+    }
+    int nextLength = length;
+    bool isDotted = false;
+    char c = toupper(*p);
+    switch (c) {
+      case 'L':
+        // set length
+        ++p;  // skip the L
+        length = (int)strtol(p, &p, 10);  // Set new length (allowed 1-64)
+        nextLength = 0;  // No note to play
+      	break;
+
+      case 'O':
+        // set octave
+        ++p;  // skip the 0
+        octave = (int)strtol(p, &p, 10);  // Set new octave (allowed 0-7)
+        nextLength = 0;  // No note to play
+      	break;
+
+      case 'T':
+        // set tempo
+        ++p;  // skip the T
+        tempo = (int)strtol(p, &p, 10);  // Set new tempo
```

Comment:
It says in the documentation that the tempo is limited to 32 to 255. There's no checking or enforcing that here.